quetz/dao.py

```diff
@@ -677,3 +678,86 @@ def get_jobs(self):
     def get_job(self, job_id: int):
         job = self.db.query(Job).filter(Job.id == job_id).one_or_none()
         return job
+
+    def incr_download_count(
+        self,
+        channel: str,
+        filename: str,
+        platform: str,
+        timestamp: Optional[datetime] = None,
+    ):
+
+        metric_name = "download"
+
+        package_version = (
+            self.db.query(PackageVersion)
+            .filter(PackageVersion.channel_name == channel)
+            .filter(PackageVersion.filename == filename)
+            .filter(PackageVersion.platform == platform)
+        ).one()
+
+        package_version.download_count += 1
+
+        q = (
+            self.db.query(PackageVersionMetric)
+            .filter(PackageVersionMetric.package_version == package_version)
+            .filter(PackageVersionMetric.metric_name == metric_name)
+        )
+
+        if timestamp is None:
+            timestamp = datetime.utcnow()
+
+        for interval in IntervalType:
+            now_interval = round_timestamp(timestamp, interval)
+            m = (
+                q.filter(PackageVersionMetric.period == interval)
+                .filter(PackageVersionMetric.timestamp == now_interval)
+                .one_or_none()
+            )
+
+            if m is None:
+                package_version = (
+                    self.db.query(PackageVersion)
+                    .filter(PackageVersion.channel_name == channel)
+                    .filter(PackageVersion.filename == filename)
+                    .filter(PackageVersion.platform == platform)
+                ).one()
+                m = PackageVersionMetric(
+                    package_version=package_version,
+                    metric_name=metric_name,
+                    period=interval,
+                    timestamp=now_interval,
+                )
+                self.db.add(m)
+                self.db.flush()
```

Comment:
do we need to flush here?

Reply:
yes, otherwise incrementing the count wouldn't work (`m.count` would be None). flush does not yet commit, only synchronises with the database engine